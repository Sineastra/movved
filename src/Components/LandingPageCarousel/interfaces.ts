export interface slideIntF {
	image: string,
	link: string
}