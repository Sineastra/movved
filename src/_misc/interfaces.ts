export interface signedUserInfoInterface {
	_id: string,
	name: string,
	email: string,
	password: string,
	profilePic: string | null,
}

export interface movieCommentIntF {
	_id: string,
	user: signedUserInfoInterface,
	comment: string
}

export interface movieInterface {
	_id: string,
	title: string,
	engTitle: string,
	genres: string[],
	smallPoster: string,
	bigPoster: string,
	isSubbed: boolean,
	isDubbed: boolean,
	year: number,
	watchLink: string,
	trailerLink: string,
	comments: movieCommentIntF[],
	actors: string[],
	director: string,
	description?: string,
	rating?: number,
	views?: number,
}

export interface filterIntF {
	name: string,
	title: string,
	options: {
		optionTitle: string,
		optionQuery: string
	}[]
}

export interface moviesForGridIntF {
	movies: movieInterface[],
	totalMoviesCount: number
}