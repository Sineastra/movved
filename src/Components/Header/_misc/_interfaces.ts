import { signedUserInfoInterface } from "../../../_misc/interfaces"
import React from "react"


export interface desktopPropsInterface {
	loggedUser: signedUserInfoInterface | null,
	className?: string,
}

export interface asideNavPropsInterface {
	loggedUser: signedUserInfoInterface | null,
	className?: string,
}

export interface mobilePropsInterface {
	toggleAsideNav: (event: React.MouseEvent) => void,
	className?: string
}