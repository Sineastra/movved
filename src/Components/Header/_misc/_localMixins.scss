@import "../../../_scssPartials/mixins/mixins";

@mixin aside-nav-link {
    height: 25%;
    text-align: start;
    text-decoration: none;
    @include centeredFlex(row);
    color: var(--theme-text-color);

    &:hover {
        color: var(--theme-secondary-color)
    }
}