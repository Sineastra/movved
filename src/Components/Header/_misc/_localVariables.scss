$aside-nav-height: 300px;
$header-height: 65px;