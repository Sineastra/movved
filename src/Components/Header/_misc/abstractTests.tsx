import { render } from "@testing-library/react"
import { BrowserRouter } from "react-router-dom"
import { ReactElement } from "react"


interface navLinksTestsIntF {
	navLinksTexts: string[],
	navLinksHrefs: string[],
	container: ReactElement
}

const runNavLinksTests = ({ navLinksTexts, navLinksHrefs, container }: navLinksTestsIntF) => {
	describe("---> Testing navigation links", () => {
		const generateData = () => {
			const searchRegex = new RegExp(navLinksTexts.join("|"), "i")
			const { getAllByText } = render(
				<BrowserRouter>
					{ container }
				</BrowserRouter>,
			)
			const navLinks = getAllByText(searchRegex)

			return navLinks
		}

		it("should show proper navigation links texts", () => {
			const navLinks = generateData()

			expect(navLinks.length).toBe(3)
		})

		it("nav links should lead to proper pages", () => {
			const navLinks = generateData()

			navLinks.forEach((x, i) => {
				expect(x.closest('a')).toHaveAttribute('href', navLinksHrefs[i])
			})
		})
	})
}

export default runNavLinksTests