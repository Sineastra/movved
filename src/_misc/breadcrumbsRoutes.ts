// because i wont be doing route obj in this app. too much refactoring needed, and i am lazy.

import MovieDetailsBreadcrumb from "../Components/Breadcrumbs/MovieDetailsBreadcrumb"


const breadcrumbsRoutes = [
	{ path: '/', breadcrumb: 'Home' },
	{ path: '/movie/:id', breadcrumb: MovieDetailsBreadcrumb },
]

export default breadcrumbsRoutes