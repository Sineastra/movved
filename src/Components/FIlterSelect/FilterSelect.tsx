import styles from "./FilterSelect.module.scss"
import { useLocation, useNavigate } from "react-router-dom"
import { BaseSyntheticEvent } from "react"
import { filterIntF } from "../../_misc/interfaces"


interface propsIntF extends filterIntF {
	ariaLabel?: string
}
const FilterSelect = ({ name, title, options, ariaLabel = '' }: propsIntF) => {
	const navigate = useNavigate()
	const queries = useLocation()

	// TODO: extract query parser if needed.
	const onChange = (e: BaseSyntheticEvent) => {
		const query = e.target.value
		const urlSearchParams = new URLSearchParams(queries.search)

		if (query) {
			urlSearchParams.set(name, query)
		} else if (name) {
			urlSearchParams.delete(name)
		}

		const areThereQueries = Array.from(urlSearchParams.entries()).length > 0

		navigate({
			search: areThereQueries ? `?${ urlSearchParams.toString() }` : '',
		})
	}

	return (
		<select name={ name } defaultValue={ title } className={ styles.select } onChange={ onChange }
		        aria-label={ ariaLabel }>
			<option value={ title } disabled>-- { title } --</option>
			<option value="">All</option>
			{ options.map(x => (
				<option value={ x.optionQuery } key={ x.optionTitle }>{ x.optionTitle }</option>
			)) }
		</select>
	)
}

export default FilterSelect