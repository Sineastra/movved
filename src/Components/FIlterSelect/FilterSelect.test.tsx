import { filterIntF } from "../../_misc/interfaces"
import renderer from "react-test-renderer"
import { BrowserRouter } from "react-router-dom"
import FilterSelect from "./FilterSelect"
import { render, screen } from "@testing-library/react"
import user from "@testing-library/user-event"
import mock = jest.mock


const mockedFilterData: filterIntF = {
	name: "name",
	title: "title",
	options: [{
		optionTitle: 'optionTitle',
		optionQuery: 'optionQuery',
	}, {
		optionTitle: 'optionTitle1',
		optionQuery: 'optionQuery1',
	}],
}

describe("---> Testing FilterSelect component", () => {
	beforeAll(() => {
		global.history.pushState({}, 'title', '/movies')
	})

	const renderScreen = () =>
		render(
			<BrowserRouter>
				<FilterSelect name={ mockedFilterData.name } title={ mockedFilterData.title }
				              options={ mockedFilterData.options }/>
			</BrowserRouter>,
		)
	it("snapshot test", () => {
		const tree = renderer.create(
			<BrowserRouter>
				<FilterSelect name={ mockedFilterData.name } title={ mockedFilterData.title }
				              options={ mockedFilterData.options }/>
			</BrowserRouter>,
		).toJSON()

		expect(tree).toMatchSnapshot()
	})
	it("should correctly add search query", async () => {
		renderScreen()

		const select = screen.getByRole("combobox")
		const option = screen.getByText<HTMLOptionElement>("optionTitle")

		await user.selectOptions(select, option)
		const search = new URLSearchParams(global.location.search)

		expect(search.get(mockedFilterData.name)).toBe("optionQuery")
	})
	it("should correctly remove search query", async () => {
		renderScreen()

		const select = screen.getByRole("combobox")
		const option = screen.getByText<HTMLOptionElement>("optionTitle")
		const resetOption = screen.getByText<HTMLOptionElement>("All")

		await user.selectOptions(select, option)
		await user.selectOptions(select, resetOption)
		const search = new URLSearchParams(global.location.search)

		expect(search.get(mockedFilterData.name)).toBe(null)
	})
	it("should not add but replace queries with the same name", async () => {
		renderScreen()

		const select = screen.getByRole("combobox")
		const option = screen.getByText<HTMLOptionElement>("optionTitle")
		const option1 = screen.getByText<HTMLOptionElement>("optionTitle1")

		await user.selectOptions(select, option)
		await user.selectOptions(select, option1)
		const search = new URLSearchParams(global.location.search)

		expect(search.get(mockedFilterData.name)).toBe("optionQuery1")
		expect(Array.from(search.entries()).length).toBe(1)
	})
})