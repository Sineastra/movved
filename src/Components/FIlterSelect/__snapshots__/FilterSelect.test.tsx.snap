// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`---> Testing FilterSelect component snapshot test 1`] = `
<select
  aria-label=""
  className="select"
  defaultValue="title"
  name="name"
  onChange={[Function]}
>
  <option
    disabled={true}
    value="title"
  >
    -- 
    title
     --
  </option>
  <option
    value=""
  >
    All
  </option>
  <option
    value="optionQuery"
  >
    optionTitle
  </option>
  <option
    value="optionQuery1"
  >
    optionTitle1
  </option>
</select>
`;
